File: java/revenj-core/src/main/java/org/revenj/patterns/OlapCubeQuery.java
package org.revenj.patterns;

import java.util.*;

public interface OlapCubeQuery<TSource extends DataSource> {
	Set<String> getDimensions();

	Set<String> getFacts();

	List<Map<String, Object>> analyze(
			List<String> dimensions,
			List<String> facts,
			Collection<Map.Entry<String, Boolean>> order,
			Specification<TSource> filter,
			Integer limit,
			Integer offset);

	default List<Map<String, Object>> analyze(
			List<String> dimensions,
			List<String> facts,
			Specification<TSource> filter) {
		return analyze(dimensions, facts, null, filter, null, null);
	}

	default List<Map<String, Object>> analyze(
			List<String> dimensions,
			List<String> facts) {
		return analyze(dimensions, facts, null, null, null, null);
	}

	default OlapCubeQueryBuilder<TSource> builder() {
		return new OlapCubeQueryBuilder(this);
	}

	class OlapCubeQueryBuilder<TSource extends DataSource> {

		private final OlapCubeQuery<TSource> query;
		private final List<String> dimensions = new ArrayList<>();
		private final List<String> facts = new ArrayList<>();
		private Integer resultLimit;
		private Integer resultOfset;
		private final Map<String, Boolean> order = new LinkedHashMap<>();

		OlapCubeQueryBuilder(OlapCubeQuery<TSource> query) {
			this.query = query;
		}

		public OlapCubeQueryBuilder<TSource> use(String dimensionOrFact) {
			if (query.getDimensions().contains(dimensionOrFact)) {
				dimensions.add(dimensionOrFact);
			} else if (query.getFacts().contains(dimensionOrFact)) {
				facts.add(dimensionOrFact);
			} else {
				throw new IllegalArgumentException("Unknown dimension or fact: " + dimensionOrFact +
						". Use getDimensions or getFacts method for available dimensions and facts");
			}
			return this;
		}

		public OlapCubeQueryBuilder<TSource> ascending(String result) {
			return orderBy(result, true);
		}

		public OlapCubeQueryBuilder<TSource> descending(String result) {
			return orderBy(result, false);
		}

		private OlapCubeQueryBuilder<TSource> orderBy(String result, boolean ascending) {
			if(!query.getDimensions().contains(result) && !query.getFacts().contains(result)) {
				throw new IllegalArgumentException("Unknown result: " + result
						+ ". Result can be only field from used dimensions and facts.");
			}
			order.put(result, ascending);
			return this;
		}

		public OlapCubeQueryBuilder<TSource> take(int limit) {
			return limit(limit);
		}

		public OlapCubeQueryBuilder<TSource> limit(int limit) {
			if (limit < 1) {
				throw new IllegalArgumentException("Invalid limit value. Limit must be positive");
			}
			resultLimit = limit;
			return this;
		}

		public OlapCubeQueryBuilder<TSource> skip(int offset) {
			return offset(offset);
		}

		public OlapCubeQueryBuilder<TSource> offset(int offset) {
			if (offset < 1) {
				throw new IllegalArgumentException("Invalid offset value. Offset must be positive");
			}
			resultOfset = offset;
			return this;
		}

		public List<Map<String, Object>> analyze() {
			return query.analyze(
					dimensions,
					facts,
					order.entrySet(),
					null,
					resultLimit,
					resultOfset
			);
		}

		public List<Map<String, Object>> analyze(Specification<TSource> specification) {
			return query.analyze(
					dimensions,
					facts,
					order.entrySet(),
					specification,
					resultLimit,
					resultOfset
			);
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/patterns/OlapCubeQuery.scala
package net.revenj.patterns

import scala.collection.mutable.ArrayBuffer
import scala.concurrent.Future

/** Olap cube is online analytical processing concept used for extracting business intelligence.
  * At it's core it's just a grouping of data by some dimensions and aggregation
  * of values through facts. Facts can be sum, count, distinct and various others concepts.
  * Cube can be made from various data sources: aggregates, snowflakes, SQL, LINQ, etc...
  *
  * DSL example:
  * {{{
  * module Finance {
  *   aggregate Payment {
  *     Timestamp createdAt { versioning; }
  *     String    account;
  *     Money     total;
  *     calculated Int year from 'it => it.Year';
  *   }
  *
  *   cube<Payment> Analysis {
  *     dimension account;
  *     dimension year;
  *     count     createdAt;
  *     sum       total;
  *   }
  * }
  * }}}
  */
trait OlapCubeQuery[T <: DataSource] {
  val dimensions: Set[String]
  val facts: Set[String]

  def analyze(
      dimensions: Seq[String],
      facts: Seq[String],
      order: Seq[(String, Boolean)] = Seq.empty,
      filter: Option[Specification[T]] = None,
      limit: Option[Int] = None,
      offset: Option[Int] = None): Future[IndexedSeq[Map[String, Any]]]

  def analyze(dimensionsAndFacts: Seq[String], filter: Specification[T]): Future[IndexedSeq[Map[String, Any]]] = {
    analyze(
      dimensionsAndFacts.filter(dimensions.contains),
      dimensionsAndFacts.filter(facts.contains),
      Seq.empty,
      Option(filter),
      None,
      None
    )
  }
}

object OlapCubeQuery {

  implicit def builder[T <: DataSource](query: OlapCubeQuery[T]): OlapCubeQueryBuilder[T] = new OlapCubeQueryBuilder(query)

  class OlapCubeQueryBuilder[T <: DataSource](query: OlapCubeQuery[T]) {
    private val dimensions = ArrayBuffer.newBuilder[String]
    private val facts = ArrayBuffer.newBuilder[String]
    private var resultLimit: Option[Int] = None
    private var resultOffset: Option[Int] = None
    private val order = ArrayBuffer.newBuilder[(String, Boolean)]

    def use(dimensionOrFact: String): this.type = {
      require(dimensionOrFact ne null, "null value provided for dimension or fact")
      require(dimensionOrFact.length != 0, "empty value provided for dimension or fact")

      if (query.dimensions.contains(dimensionOrFact)) {
        dimensions += dimensionOrFact
      } else if (query.facts.contains(dimensionOrFact)) {
        facts += dimensionOrFact
      } else {
        throw new IllegalArgumentException("Unknown dimension or fact: " + dimensionOrFact + ". Use dimensions or facts method for available dimensions and facts")
      }
      this
    }

    def ascending(result: String): OlapCubeQueryBuilder[T] = orderBy(result, ascending = true)

    def descending(result: String): OlapCubeQueryBuilder[T] = orderBy(result, ascending = false)

    private def orderBy(result: String, ascending: Boolean): this.type = {
      require(query.dimensions.contains(result) || query.facts.contains(result), "Unknown result: " + result + ". Result can be only field from used dimensions and facts.")
      order += result -> ascending
      this
    }

    def take(count: Int) = limit(count)

    def limit(count: Int): this.type = {
      require(count > 0, "Invalid limit value. Limit must be positive")
      resultLimit = Some(count)
      this
    }

    def drop(count: Int) = offset(count)

    def offset(count: Int): this.type = {
      require(count > 0, "Invalid offset value. Offset must be positive")
      resultOffset = Some(count)
      this
    }

    def analyze(specification: Option[Specification[T]] = None): Future[IndexedSeq[Map[String, Any]]] = {
      query.analyze(dimensions.result(), facts.result(), order.result(), specification, resultLimit, resultOffset)
    }
  }
}
